### PR TITLE
Fix/paid subscribers on self hosted

### LINKED ISSUE #30986: Newsletter: "undefined" displayed instead of subscribers count
#### Steps to reproduce

1. Start with a site running Jetpack trunk and connected to WordPress.com.
2. Go to Pages > add New, create a new subscribe page with a subscribe block, and add a payment plan there.
3. Go to Posts > Add New, and create a new post.

You'll see the following (note the "undefined") displayed in the different subscriptions panels:

Screenshot 2023-05-26 at 12 39 52
Screenshot 2023-05-26 at 12 38 46

### PR DESCRIPTION
**Pull request by millerf**
Fixes https://github.com/Automattic/jetpack/issues/30986

## Proposed changes:
* Merges [PR #31094](https://github.com/Automattic/jetpack/pull/31094) and [PR #31009](https://github.com/Automattic/jetpack/pull/31009)
* Create a proper test instructions based on @n3f 's previous work

### Other information:

- [X] Have you written new tests for your changes, if applicable?
- [X] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [X] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?

Nope

## Testing instructions:

* Create a jurassic ninja with `https://jurassic.ninja/create?jetpack-beta&branches.jetpack=fix/paid-subscribers-on-self-hosted&wp-debug-log`
* Push this code on your sandbox
```
bin/jetpack-downloader test jetpack fix/paid-subscribers-on-self-hosted
bin/jetpack-downloader test jetpack-mu-wpcom-plugin fix/paid-subscribers-on-self-hosted

```
* Point your newly created machine to your sandbox from https://respectable-horseshoe.jurassic.ninja/wp-admin/options-general.php?page=companion_settings
`{USER}.SANBOX_URL`
* On your sandbox, put your sandbox in sandbox mode and sandbox the async job counting the paid subscribers
```
function sandbox_async_jobs( $sandboxed, $type ) {
	return in_array( $type, array(
		'wpcom_subs_update_total_paid_subscribers',
	) );
}

[...]

define( 'USE_STORE_SANDBOX', true );
```
* Start the watcher `$ ./async-jobs/watcher.php`
* Open your sandbox to the outside world `sudo global-http enable`
* Congrats! Now, your ninja instance will call your sandbox.
* Set up jetpack (on Free plan is enough)
* Enable subscriptions in https://condemned-marmot.jurassic.ninja/wp-admin/admin.php?page=jetpack#/discussion
Screenshot 2023-06-01 at 16 47 13

* Create a new post. 
* Go to Earn via the Newsletter Access panel and set stripe and a paid plan
Screenshot 2023-06-01 at 16 48 25

#### Ready to test
* Create a new post
* Notice the 0 and NO `undefined` in the panel

Screenshot 2023-06-01 at 16 49 28

* Create a subscribe block and publish the post. https://condemned-marmot.jurassic.ninja/2023/06/01/9/
* Open a new incognito window and pay the subscription
Screenshot 2023-06-01 at 16 51 52

* With test card it is fine
Screenshot 2023-06-01 at 16 52 31

* Notice the async-job was launched
Screenshot 2023-06-01 at 16 53 59

* Now notice a new post has one more paid-subscriber 🥳 
Screenshot 2023-06-01 at 17 04 30

* If you see `0`. It could be that the production job has overwritten the sandbox job
* You can re-write it with
```
~/public_html (trunk) $ wpsh
> print(get_blog_id_from_url('condemned-marmot.jurassic.ninja'))
219683543
> $blog_id=219683543
> require('async-jobs/includes/wpcom-subscriptions.php')
> jobs_wpcom_subs_update_total_paid_subscribers((object)['data'=>['blog_id'=>$blog_id]])
```
* And check:
```
 > gpr SELECT * FROM subscription_counts WHERE blog_id=219683543
+-----------------------+-------------------+------------+-----------+---------+--------------------+
| subscription_count_id | subscription_type | user_type  | blog_id   | post_id | subscription_count |
+-----------------------+-------------------+------------+-----------+---------+--------------------+
| 174889426             | posts             | wpcom      | 219683543 |         | 2                  |
| 174889427             | posts             | wpcom      | 219683543 |         | 2                  |
| 174889740             | posts             | wpcom_paid | 219683543 |         | 1                  |
+-----------------------+-------------------+------------+-----------+---------+--------------------+

```
* Now reload the new post page and it should be good

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 1]
```diff
@@ -173,7 +173,7 @@ function fetch_subscriber_counts() {
 	$subs_count = 0;
 	if ( is_wpcom() ) {
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:174]: Here's some optional feedback

```suggestion
	if ( is_wpcom() ) {
	         $include_paid_subscribers = true;
```

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 2]
```diff
@@ -173,7 +173,7 @@ function fetch_subscriber_counts() {
...
 		$subs_count = array(
-			'value' => \wpcom_fetch_subs_counts( false ),
+			'value' => \wpcom_fetch_subs_counts( true ),
```
Comment (romarioraffington) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:176]:
```suggestion
			'value' => \wpcom_fetch_subs_counts( $include_paid_subscribers ),
```

File: projects/plugins/jetpack/changelog/earn-newsletters-fix-undefined-counts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: bugfix
+
+Fixes the check for falsy values in the `getReachForAccessLevelKey` function
```
Comment (n3f) [inline projects/plugins/jetpack/changelog/earn-newsletters-fix-undefined-counts]: Let's update this changelog...

**Comment by n3f** [comment]
Change needed to support atomic sites is here: D112670-code ✅ (deployed).

**Review by n3f — APPROVED** [review]
WFM. 👍

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 2 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:176]: We gonna remove that parameter anyway. It is not useful anymore

Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:176]: Do we need to worry about the potential impact on performance this change may have on WordPress.com?

**Review by jeherve — APPROVED** [review]
This seems to test well for me. I'm always a bit worried by the potential impacts on performance these additional queries may have. But you're more familiar with this part of the codebase than I am, so approving. :)

Edit: ha, I'm late with my review anyway :)

**Comment by anomiex** [comment]
I'm not feeling very confident in cherry-picking this on the Friday of code-freeze week when there are performance concerns with the change. I recommend that next time you aim for earlier in the week.

But I suppose the performance concerns are only with code that runs only on wpcom, so (1) if this blows up it will hopefully happen when the daily deploy does, and (2) if that part of it does need reverting once 12.2 goes out, that will only be needed on wpcom too. The rest of the PR consists of JS changes that should continue to function (although incorrectly showing 0 for the number of paid subscribers) even if the wpcom part is reverted.

Cherry-picked in 970789c600.

**Pull request merged**